### PR TITLE
p2p: add eviction metrics and cleanup dialing error handling

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: internal/p2p/peermanager.go [hunk 1]
```diff
@@ -513,12 +513,21 @@ func (m *PeerManager) HasDialedMaxPeers() bool {
...
+			return NodeAddress{}, ctx.Err()
+		}
+
+		if !m.HasDialedMaxPeers() {
+			continue
+		}
+
 		select {
 		case <-m.dialWaker.Sleep():
+			continue
```
Comment (creachadair) [inline internal/p2p/peermanager.go:530]: Fine, but this is what it does anyway.

**Review by creachadair — APPROVED** [review]

File: internal/p2p/peermanager.go [hunk 2]
```diff
@@ -513,12 +513,21 @@ func (m *PeerManager) HasDialedMaxPeers() bool {
...
+			return address, nil
 		}
+
+		if ctx.Err() != nil {
+			return NodeAddress{}, ctx.Err()
+		}
+
+		if !m.HasDialedMaxPeers() {
+			continue
+		}
```
Comment (cmwaters) [inline internal/p2p/peermanager.go:526]: We only get here if there are no addresses to dial, so having this if statement means we will just be constantly spinning in a for loop

File: internal/p2p/router.go [hunk 3]
```diff
@@ -642,6 +639,7 @@ func (r *Router) connectPeer(ctx context.Context, address NodeAddress) {
 
 	if err := r.runWithPeerMutex(func() error { return r.peerManager.Dialed(address) }); err != nil {
 		r.logger.Error("failed to dial peer", "op", "outgoing/dialing", "peer", address.NodeID, "err", err)
+		r.peerManager.dialWaker.Wake()
```
[the remark this answers is not in the document]
Comment (tychoish) [reply internal/p2p/router.go:642]: I think it's a problem if we fail to connect to a peer and *don't* call the wakeup thread, this is a case where we would otherwise shed peers and not push a new peer for dialing

File: internal/p2p/peermanager.go (hunk 1 above)
Comment (tychoish) [reply internal/p2p/peermanager.go:530]: yeah, I was just trying to make the paths through these functions a bit clearer.

File: internal/p2p/peermanager.go (hunk 2 above)
Comment (cmwaters) [reply internal/p2p/peermanager.go:526]: You mean if connection slots are full? It's that and if there are no other peers to dial. I believe the point of the sleeper was basically there's no point continually calling `TryDialNext` until we either have a new peer through PEX, a peer disconnects or we're able to redial a peer so let's wait until then

Comment (tychoish) [reply internal/p2p/peermanager.go:526]: we're definitely seeing nodes gradually shed peer connections and just never try to dial more peers. I think the "avoid dialing peers unless we need to" is being really careful with a relatively inexpensive operation (dialing) at the expense of being able to maintain connections to enough peers to make progress.

Comment (cmwaters) [reply internal/p2p/peermanager.go:526]: I understand you hypothesis that `DialNext` might be somehow blocking and thus we stop attempting to dial peers. I'm just not sure that the `dialWaker` is the culprit. 

> "avoid dialing peers unless we need to"

It's not so much this, but "don't look for a new peer to dial, if we know we don't have any"

Comment (cmwaters) [reply internal/p2p/peermanager.go:526]: One thing we can try, is to just remove the `dialWaker` entirely and see if that changes anything

Comment (tychoish) [reply internal/p2p/peermanager.go:526]: I'd definitely be down to do that as an experiment. I think this PR is worth merging, and then following up with removing that. 

Frankly I think the larger problem is that we'll try very hard to hit the peers at the top of the ranked list, and struggle to get peers because of our bias to the sort order.

Comment (cmwaters) [reply internal/p2p/peermanager.go:526]: > Frankly I think the larger problem is that we'll try very hard to hit the peers at the top of the ranked list, and struggle to get peers because of our bias to the sort order.

I don't think this is too much of a problem because we apply exponential backoff to peers that we failed to dial, and their score is temporarily reduced

Comment (tychoish) [reply internal/p2p/peermanager.go:526]: I think the other PR would mean that we would dial more and this is mostly just cleanup.

Comment (cmwaters) [reply internal/p2p/peermanager.go:526]: I don't think this will dial any more unless we had a bug beforehand where we weren't causing `Wake` in a situation where a new dialable peer pops up

Comment (tychoish) [reply internal/p2p/peermanager.go:526]: There is the one case where a failure to dial *wouldn't* lead to us dialing new peers. 

I like the idea of seeing if we can pull out of needing the dialWaker at all.

Comment (cmwaters) [reply internal/p2p/peermanager.go:526]: > There is the one case where a failure to dial wouldn't lead to us dialing new peers.

Where is this? As I said, if we can just fix the cases where we actually need to `Wake` instead of falling back to calling `Wake` at a random interval then I think this is a far better solution

Comment (tychoish) [reply internal/p2p/peermanager.go:526]: I removed this and added what I think is the missing wake call.

**Review by cmwaters — APPROVED** [review]
LGTM

File: internal/p2p/router.go [hunk 4]
```diff
@@ -591,9 +584,8 @@ LOOP:
 		switch {
 		case errors.Is(err, context.Canceled):
 			break LOOP
-		case err != nil:
-			r.logger.Error("failed to find next peer to dial", "err", err)
-			break LOOP
+		case address == NodeAddress{}:
+			continue LOOP
```
Comment (cmwaters) [inline internal/p2p/router.go:588]: This only happens when the context is cancelled right?

Comment (cmwaters) [reply internal/p2p/router.go:588]: i.e. `DialNext` either always returns a node address or returns and empty one and `ctx.Err()`

Comment (tychoish) [reply internal/p2p/router.go:588]: yeah, that's the case, but even if it weren't we shouldn't abort dialing peers just because we had a problem getting an error from dialnext: until the context is canceled we should keep  trying to dial.

File: internal/p2p/router.go [hunk 5]
```diff
@@ -603,7 +595,7 @@ LOOP:
 			// create connections too quickly.
 
 			r.dialSleep(ctx)
-			continue
+			continue LOOP
```
Comment (williambanfield) [inline internal/p2p/router.go:598]: Why this change?

File: internal/p2p/router.go [hunk 6]
```diff
@@ -642,6 +634,7 @@ func (r *Router) connectPeer(ctx context.Context, address NodeAddress) {
 
 	if err := r.runWithPeerMutex(func() error { return r.peerManager.Dialed(address) }); err != nil {
 		r.logger.Error("failed to dial peer", "op", "outgoing/dialing", "peer", address.NodeID, "err", err)
+		r.peerManager.dialWaker.Wake()
```
Comment (williambanfield) [inline internal/p2p/router.go:637]: Why don't we call `DialFailed` here?

Comment (tychoish) [reply internal/p2p/router.go:637]: because the dialing succeeded we just throw out the connection because this is your race bug.

File: internal/p2p/router.go (hunk 4 above)
Comment (tychoish) [reply internal/p2p/router.go:588]: I see, I still think it's reasonable to have the code be a little defensive here.

File: internal/p2p/router.go (hunk 6 above)
Comment (williambanfield) [reply internal/p2p/router.go:637]: I think we should wake the dial sleeper from in the `Dialed` method for the errors where we expect to need to start a new dial, namely:

https://github.com/tendermint/tendermint/blob/83526cacbcd998b33d2669088a688c7e492da8ff/internal/p2p/peermanager.go#L672

and 

https://github.com/tendermint/tendermint/blob/83526cacbcd998b33d2669088a688c7e492da8ff/internal/p2p/peermanager.go#L675
